Few more updates

kdmurthy opened the pull request:
feat: Added id_whitelist and class_whitelist options
feat: Added root_node option to get relative CSS selector
feat: Added a `get_options` option. Setting get_options to a function that accepts an element as a parameter invokes the function for updating the options before the selector is generated. In effect, options become dynamic per element.
refactor: Polyfill `Array.find` is moved from the test to src module.
refactor: Removed polyfill `Array.find`
updated: README.md

@fczbkk I need the root_node option for one of my projects. Since, the option doesn't change the original behaviour, I will be grateful if it is kept :). Otherwise, I need to maintain a fork.

fflorent commented:
Thanks @kdmurthy ! 

I am thinking about the behavior of whitelist and blacklist. I wonder if we couldn't also ask the user to provide a function which would be given the class lists / ids / attributes and would return the values that have to be in the selector (in the order the user wants).

Let's say:
```javascript
var custom_options = {
  selectors: ['tag', 'id', 'class'],
  /**
  classes(classList) {
    return classList.filter(className => className.startWith('foo')).sort();
  },
  ids(id) {
    return id === 'unwanted' ? null : id;
  },
  attrs(attrList) {
    return attrList.filter([key, value] => value && /-wanted-/.test(key));
  }
};
```

Maybe that could be complementary. If you think that would be a relevant contribution, I can submit implement that in another PR.

Cheers!
Florent

kdmurthy commented:
I toyed around with the idea of accepting a `function` in the white/black lists and then thought it will complicate the usage. The `function` accepts an element and value and returns a boolean. Something like:
```js
var input_attr = function(e, attr) {
  if(attr != 'type' && attr != 'name') return false;
  if(e.tagName.toLower() == 'input') {
     if(e.getAttribute('type') == 'radio' || e.getAttribute('type') == 'checkbox') {
        return true ;
     }
     return attr == 'name' ;
  }
}

attribute_whitelist: [ input_attr ]
```

Then I thought I am complicating the usage and dropped the idea :)

fflorent commented:
Well, the idea behind is not about giving a function in the whitelists or in the blacklists but giving the user the ability to accept / refuse the attributes using a function.

Thus a user may filter the classes / attributes / ids they want using either these functions or whitelists.

Both would be equivalent:
```javascript
const my_selector_generator = new CssSelectorGenerator({
  selectors: ['attribute'],
  attributes(attrList, element) {
    return attrList.filter(([key, value]) => /^wanted-/.test(key));
  }
});

const my_selector_generator = new CssSelectorGenerator({
  selectors: ['attribute'],
  attribute_blacklist: [/.*/],
  attribute_whitelist: [/^wanted-/]
});
```

Except that you may do more advanced stuffs using the first form (filtering depending on the value or the element).

I don't feel the need right now to have this, so I won't be really bothered if this idea isn't accepted.

Florent

kdmurthy commented:
What happens if both `attributes` method as well as `attribute_blacklist` is given. Once you sort the priority should be OK. I guess we should check the `whitelist` first, followed by the `attributes` method and then `blacklist`. I suggest you post a pull request and see @fczbkk feels about it :)

fflorent commented:
>Once you sort the priority should be OK. I guess we should check the whitelist first, followed by the attributes method and then blacklist.

I suggest that if `attributes` is given, whitelists and blacklists are ignored (because they would be redundant).

kdmurthy commented:
Not very sure.
1. A rule of API design is that if you allow a user to provide a parameter, the parameter should be used.
2. It is much easier to add an entry into blacklist/whitelist than modifying a method.

kdmurthy commented:
@fflorent I merged your changes (including renaming `notInList` to `inList` - what was I thinking?). I also added support for nth-of-type selector. Especially useful when creating relative selectors for a table. Updated README and tests. Added a complex test to check with `nthtype` instead of `nthchild`.
@fczbkk If we set selectors to only 'nthchild' the `should pass complex test` fails (4 failures). I could not figure out why. The tests pass if 'prefix_tag' is set to true.

fflorent commented:
Great, thanks @kdmurthy!

fflorent commented on src/css-selector-generator.coffee:
```diff
@@ -102,23 +110,32 @@ class CssSelectorGenerator
+      return x.test item
+
+  itemMatches: (item, list) ->
+    return list.find (x) ->
```
Could you use `.some` in order to support browsers before ES2015 ?

kdmurthy commented:
We need the return value, so we can't. However, I use a `polyfill` for `Array.find`. We actually have it in the test sources, we can move it to main sources. I will update the pull request soon.

fflorent commented:
I wouldn't be confortable with injecting the polyfill in this library without the consent of the user (who may have their own polyfill). We don't know the impacts of this.

Rather, I propose: 
```coffeescript
  itemMatches: (searchedItem, list) ->
    for item in list
      if typeof searchedItem == 'string'
        return item if searchedItem == item
      else
        return item if searchedItem.test item
    return null
```

fczbkk commented:
Hey guys. Thank you for the discussion and pull request. Few things:

- Please do not cram several changes and features into single pull request. I do not have the capacity to review it all at once. @kdmurthy  please split the changes into single-purpose pull requests.
- Adding a polyfill into a library is bad practice. I agree with @fflorent. I have seen polyfills collide and backfire horribly. I do not think external library should mutate native objects in any way.

kdmurthy commented:
@fczbkk Not good with github - the UI allows me to create a single pull request. I will check whether I can separate them. That said, I ensured that all existing tests pass without changes and added new tests for each of the feature.
Agree about polyfill - though in practice most of them are from MDN and do not really clash. :).

fczbkk commented:
@kdmurthy Seems like you are creating pull request from single branch ("master"). Create a standalone branch for each feature/change. Then create pull request from that branch. You can create any number of pull requests.

More info at https://help.github.com/articles/creating-a-pull-request-from-a-fork/

fflorent commented:
@fczbkk @kdmurthy I opened other PR and reopened the main patch of #31 in #34. If it alright for you?

Florent

kdmurthy commented:
@fflorent anything is OK with me as long as all changes go through :). Currently no time to look in detail. Later.
